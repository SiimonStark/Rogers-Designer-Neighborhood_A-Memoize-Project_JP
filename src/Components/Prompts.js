import React, { Component } from 'react';

class Prompts extends Component {
  constructor(props) {
    super(props)
    this.state = {
      prompts: [
        {
          question: "# 1: Grid the field 2x2, for 4 total squares",
          answers: {
            display: "grid",
            gridTemplateAreas: `"square-1 square-2" "square-3 square-4"`
          },
          correctAns: false
        },
        {
          question: "# 2: Align the Roads to the edge of the grid",
          answers: {
            display: "grid",
            gridTemplateAreas: `"square-1 square-2" "square-3 square-4"`
          },
          correctAns: false
        },
        {
          question:
            "# 3: Give space around the trees & Align them to the left of each grid",
          answers: {
            display: "grid",
            gridTemplateAreas: `"square-1 square-2" "square-3 square-4"`
          },
          correctAns: false
        },
        {
          question: "# 4: Center the house in the grid square",
          answers: {
            display: "grid",
            gridTemplateAreas: `"square-1 square-2" "square-3 square-4"`
          },
          correctAns: false
        },
        {
          question: "No more Prompts!!",
          answers: undefined,
          correctAns: undefined
        }
      ],
      currentPrompt: 0,
      showButton: true
    };
  }
  checkPrompt = () => {
    let index = this.state.currentPrompt;

    switch(index) {
      case 0: this.compareGrid(index)
        break;
      case 1: this.compareFlex(index)
        break;
      case 2: this.compareFlex(index)
        break;
      case 3: this.compareFlex(index)
        break;
      case 4: 
        alert('More Coming Soon')
        this.setState({showButton: false})
        break;
      default: console.log('ERROR: Switch Statement');
    }
  }
  compareGrid = (index) => {
    let newPrompts = this.state.prompts;

    if (this.props.playerInputs[0].gridInputs['gridField0']) {
      newPrompts[index].correctAns = true;
      this.setState({ prompts: newPrompts, currentPrompt: 1 }
        , this.updateCorrectPrompt())
    }
  }
  compareFlex = (index) => {
    let newPrompts = this.state.prompts;

    if (this.props.playerInputs[1].flexInputs['flexField0']) {
      newPrompts[index].correctAns = true;
      index ++
      this.setState(
        { prompts: newPrompts, currentPrompt: index }
        , this.updateCorrectPrompt()
      );
    }
  }
  updateCorrectPrompt = () => {
    let promptsAns = this.state.prompts.map(prompt => prompt.correctAns);
    this.props.updatePromptAns(promptsAns);
  }
  render() {
    let button;
    this.state.showButton
      ? button = 
        <button 
          id="Prompt-check" 
          onClick={this.checkPrompt}>
          Check
        </button>
      : button = null;
    
    return (
      <footer className="Prompts">
        <article>
          <h3>Prompt Instructions:</h3>
          <p className="question">
            {this.state.prompts[this.state.currentPrompt].question}
          </p>
        </article>
        {button}
      </footer>
    );
  }
}

export default Prompts;