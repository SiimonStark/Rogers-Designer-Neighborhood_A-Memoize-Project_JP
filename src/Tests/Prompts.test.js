import React from 'react';
import Prompts from '../Components/Prompts';
import { shallow } from 'enzyme';

let updatePromptAnsMock = jest.fn();
let mockPlayerInputs = [
  {gridInputs: []},
  {flexInputs: []}
];
let mockPrompts = [
  {
    question: "# 1: Grid the field 2x2, for 4 total squares",
    answers: {
      display: "grid",
      gridTemplateAreas: `"square-1 square-2" "square-3 square-4"`
    },
    correctAns: false
  },
  {
    question: "# 2: Align the Roads to the edge of the grid",
    answers: {
      display: "grid",
      gridTemplateAreas: `"square-1 square-2" "square-3 square-4"`
    },
    correctAns: false
  },
  {
    question:
      "# 3: Give space around the trees & Align them to the left of each grid",
    answers: {
      display: "grid",
      gridTemplateAreas: `"square-1 square-2" "square-3 square-4"`
    },
    correctAns: false
  },
  {
    question: "# 4: Center the house in the grid square",
    answers: {
      display: "grid",
      gridTemplateAreas: `"square-1 square-2" "square-3 square-4"`
    },
    correctAns: false
  },
  {
    question: "No more Prompts!!",
    answers: undefined,
    correctAns: undefined
  }
];

describe('Prompts', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Prompts
        playerInputs={mockPlayerInputs}
        updatePromptAns={updatePromptAnsMock}
      />
    );
  });

  it("should match Snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match conditional Snapshot', () => {
    wrapper.state({showButtow: false})
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default state', () => {
    expect(wrapper.state()).toEqual({
      prompts: mockPrompts,
      currentPrompt: 0,
      showButton: true
    });
  });

  it('should onClick of the button invoke checkPrompts for Grid', () => {
    mockPlayerInputs[0].gridInputs.push(
      {gridField0:{gridRule: 'hello', gridValue: 'world'}}
      );
    let wrapper = shallow(
      <Prompts
      playerInputs={mockPlayerInputs}
      updatePromptAns={updatePromptAnsMock}
      />
    ).renderProp('render')();
    let mockIndex = wrapper.state('currentPrompt');
    console.log('new inputs', mockPlayerInputs);
    console.log('currPrompt', wrapper.state('currentPrompt'))
    
    wrapper.find("#Promptcheck").simulate('click');
    console.log(wrapper.state())
    wrapper.instance().compareGrid(wrapper.state('currentPrompt'));
    let updatedMockPrompts = mockPrompts;
    updatedMockPrompts[0].correctAns = true;
    // ? console.log('updated mockPrompts', updatedMockPrompts);
    expect(wrapper.state('prompts')).toEqual(updatedMockPrompts);
  });

  it.skip('should onClick of the button invoke checkPrompts for Flex', () => {
    wrapper.state({currentPrompt: 1})
    let mockIndex = wrapper.state('currentPrompt');
    mockPlayerInputs[1].flexInputs.push(
        {flexField0:{flexRule: '', flexValue: ''}}
      );
    console.log('test', mockPlayerInputs);

    wrapper.instance().checkPrompt();
    wrapper.instance().compareFlex(mockIndex);
    expect(wrapper.state()).toEqual({
      prompts: [
        {
          question: "# 2: Align the Roads to the edge of the grid",
          answers: {
            display: "grid",
            gridTemplateAreas: `"square-1 square-2" "square-3 square-4"`
          },
          correctAns: true
        }
      ]
    });
  });

  it('should invoke updateCorrectPrompts() if compareGrid() has been invoked', () => {
    let mockIndex = wrapper.state("currentPrompt");
    wrapper.instance().compareGrid(mockIndex);
    wrapper.instance().updateCorrectPrompt();
    expect(updatePromptAnsMock).toHaveBeenCalled();
  });
});