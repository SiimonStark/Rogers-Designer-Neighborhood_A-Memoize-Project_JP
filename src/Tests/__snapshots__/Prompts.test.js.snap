// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prompts should match Snapshot 1`] = `
<footer
  className="Prompts"
>
  <article>
    <h3>
      Prompt Instructions:
    </h3>
    <p
      className="question"
    >
      # 1: Grid the field 2x2, for 4 total squares
    </p>
  </article>
  <button
    id="Prompt-check"
    onClick={[Function]}
  >
    Check
  </button>
</footer>
`;

exports[`Prompts should match conditional Snapshot 1`] = `
<footer
  className="Prompts"
>
  <article>
    <h3>
      Prompt Instructions:
    </h3>
    <p
      className="question"
    >
      # 1: Grid the field 2x2, for 4 total squares
    </p>
  </article>
  <button
    id="Prompt-check"
    onClick={[Function]}
  >
    Check
  </button>
</footer>
`;
